client/consensus/grandpa/environment.go

```diff
@@ -0,0 +1,466 @@
+	}
+
+	return CompletedRounds[H, N]{
+		rounds,
+		setId,
+		voterIds,
+	}
+}
+
+// Last Returns the last (latest) completed round
+func (compRounds *CompletedRounds[H, N]) Last() CompletedRound[H, N] {
+	if len(compRounds.Rounds) == 0 {
+		panic("inner is never empty; always contains at least genesis; qed")
+	}
+	return compRounds.Rounds[0]
+}
+
+// Push a new completed round, oldest round is evicted if number of rounds
+// is higher than `NUM_LAST_COMPLETED_ROUNDS`.
+func (compRounds *CompletedRounds[H, N]) Push(completedRound CompletedRound[H, N]) {
+	// TODO they use reverse, double check this
+	idx, found := slices.BinarySearchFunc(
+		compRounds.Rounds,
+		N(completedRound.Number),
+		func(a CompletedRound[H, N], b N) int {
+			switch {
+			case N(a.Number) == b:
+				return 0
+			case N(a.Number) > b:
+				return 1
+			case N(a.Number) < b:
+				return -1
+			default:
+				panic("huh?")
+			}
+		},
+	)
+
+	if found {
+		compRounds.Rounds[idx] = completedRound
+	} else {
+		if len(compRounds.Rounds) <= idx {
+			compRounds.Rounds = append(compRounds.Rounds, completedRound)
+		} else {
+			compRounds.Rounds = append(compRounds.Rounds[:idx+1], compRounds.Rounds[idx:]...)
+			compRounds.Rounds[idx] = completedRound
+		}
+	}
+
+	if len(compRounds.Rounds) > NumLastCompletedRounds {
+		compRounds.Rounds = compRounds.Rounds[:len(compRounds.Rounds)-1]
+	}
+}
+
+// CurrentRounds A map with voter status information for currently live rounds,
+// which votes have we cast and what are they.
+type CurrentRounds[H comparable, N constraints.Unsigned] map[uint64]HasVoted[H, N]
+
+// VoterSetState The state of the current voter set, whether it is currently active or not
+// and information related to the previously completed rounds. Current round
+// voting status is used when restarting the voter, i.e. it will re-use the
+// previous votes for a given round if appropriate (same round and same local
+// key).
+type VoterSetState[H comparable, N constraints.Unsigned] scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (tve *VoterSetState[H, N]) Set(val scale.VaryingDataTypeValue) (err error) { //skipcq: GO-W1029
+	vdt := scale.VaryingDataType(*tve)
+	err = vdt.Set(val)
+	if err != nil {
+		return err
+	}
+	*tve = VoterSetState[H, N](vdt)
+	return nil
+}
+
+// Value will return the value from the underlying VaryingDataType
+func (tve *VoterSetState[H, N]) Value() (val scale.VaryingDataTypeValue, err error) { //skipcq: GO-W1029
+	vdt := scale.VaryingDataType(*tve)
+	return vdt.Value()
+}
+
+// New ...
+func (tve VoterSetState[H, N]) New() VoterSetState[H, N] {
+	vdt, err := scale.NewVaryingDataType(Live[H, N]{
+		CompletedRounds: CompletedRounds[H, N]{
+			Rounds: make([]CompletedRound[H, N], 0, NumLastCompletedRounds),
+		},
+		CurrentRounds: make(map[uint64]HasVoted[H, N]), // init the map
+	}, Paused[H, N]{})
+	if err != nil {
+		panic(err)
+	}
+	vs := VoterSetState[H, N](vdt)
+	return vs
+}
```

Comment:
I think I'm going to make `scale.VaryingDataType` an interface.  That way it's not so cumbersome to create these enums.